src/simsopt/solve/permanent_magnet_optimization.py

```diff
@@ -0,0 +1,415 @@
+    ndipoles = len(m) // 3
+    mmax_vec = np.array([mmax, mmax, mmax]).T
+    m_normalized = (np.abs(m).reshape(ndipoles, 3) / mmax_vec).reshape(ndipoles * 3)
+    # in principle, hard threshold should be sqrt(2 * reg_l0 * nu) but can always renormalize everything
+    return m * (m_normalized > 2 * reg_l0 * nu)
+
+
+def prox_l1(m, mmax, reg_l1, nu):
+    """
+    Proximal operator for L1 regularization.
+    Note that the m values are normalized before soft
+    thresholding to avoid only truncating the magnets on the
+    inner side of the configuration (which are often important!).
+
+    Args:
+        m: 2D numpy array, shape (ndipoles, 3)
+            The permanent magnet dipole vectors.
+        mmax: 1D numpy array, shape (ndipoles)
+            The maximal dipole strengths of each of the permanent magnets.
+        reg_l1: double
+            The amount of L1 regularization used in the problem.
+        nu: double
+            The strength of the "relaxing" term in the relax-and-split algorithm
+            used for permanent magnet optimization.
+    """
+    ndipoles = len(m) // 3
+    mmax_vec = np.array([mmax, mmax, mmax]).T
+    m_normalized = (np.abs(m).reshape(ndipoles, 3) / mmax_vec).reshape(ndipoles * 3)
+    return np.sign(m) * np.maximum(np.abs(m_normalized) - reg_l1 * nu, 0) * np.ravel(mmax_vec)
+
+
+def projection_L2_balls(x, mmax):
+    """
+    Project the vector x onto a series of L2 balls in R3.
+    Only used here for checking if the initial guess for the
+    permanent magnets is inside the feasible region
+    before optimization begins.
+
+    Args:
+        x: 1D numpy array, shape (ndipoles * 3)
+            The current solution vector for the dipole vectors of the magnets.
+        mmax: 1D numpy array, shape (ndipoles)
+            The maximal dipole strength of each of the magnets.
+    """
+    N = len(x) // 3
+    x_shaped = x.reshape(N, 3)
+    denom_fac = np.sqrt(np.sum(x_shaped ** 2, axis=-1)) / mmax
+    denom = np.maximum(np.ones(len(denom_fac)), denom_fac)
+    return np.divide(x_shaped, np.array([denom, denom, denom]).T).reshape(3 * N)
+
+
+def setup_initial_condition(pm_opt, m0=None):
+    """
+    If an initial guess for the dipole moments is specified,
+    checks the initial condition lies in the allowed hypersurface.
+    If an initial guess is not specified, defaults to initializing
+    the permanent magnet dipole moments to all zeros.
+
+    Args:
+        pm_opt: PermanentMagnetGrid class object
+            Permanent magnet grid for optimization.
+        m0: 1D numpy array, shape (ndipoles * 3)
+            Initial guess for the dipole vectors.
+    """
+    # Initialize initial guess for the dipole strengths
+    if m0 is not None:
+        if len(m0) != pm_opt.ndipoles * 3:
+            raise ValueError(
+                'Initial dipole guess is incorrect shape --'
+                ' guess must be 1D with shape (ndipoles * 3).'
+            )
+        m0_temp = projection_L2_balls(
```

Comment:
One typically breaks arguments into subsequent lines when there are too many arguments, or the line goes past 80 chars. It is distracting for small functions.